tools/slo-generator/slo_generator/report.py

```diff
@@ -0,0 +1,274 @@
+# Copyright 2019 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#            http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+`report.py`
+Report utilities.
+"""
+
+import logging
+from dataclasses import dataclass, asdict, fields
+from slo_generator import utils
+
+LOGGER = logging.getLogger(__name__)
+
+
+@dataclass(init=False)
+class SLOReport:
+    """SLO report dataclass. Compute an SLO report out of an SLO config and an
+    Error Budget Policy step.
+
+    Args:
+        config (dict): SLO configuration.
+        step (dict): Error budget policy step configuration.
+        timestamp (int): Timestamp.
+        client (obj): Existing backend client.
+        delete (bool): Backend delete action.
+    """
+    # pylint: disable=too-many-instance-attributes
```

Comment:
Consider refactoring, this lint is here for a reason.

For example, could we instead have references to a `SLO` class and an `ErrorBudget` class which contain their respective data?

Reply:
Refactoring is not trivial in this case, since many methods in this class are using fields from the error budget policy step. It would significantly complexify the code, and "Simple is better than complex" (Python zen).

Even if we split the Error budget fields into their own class, the `Pylint` limit here is 7 instance attributes, so that would still not be enough for both the `SLOReport` and `ErrorBudget` classes. 

IMO the linter should have the last word here. See [here](https://stackoverflow.com/questions/24434510/how-to-deal-with-pylints-too-many-instance-attributes-message) for reference.